node/src/utils/round_robin.rs

```diff
@@ -58,6 +61,14 @@ impl<I> QueueState<I> {
         }
     }
 
+    /// Remove all events from a queue.
+    async fn drain(&self) -> Vec<I> {
```

Comment:
Could this return an iterator instead?

Reply:
An iterator would block the queue permanently while it exists.

```
1. LOCK QUEUE
2. DRAIN ITEMS (<--- iterator would replace this step and would need to hold the lock)
3. UNLOCK QUEUE
```

is what is happening here.

If we returned an iterator, we'd essentially hold the lock as long as it is being iterated, and we would need a guard type to do so. I'm not sure this is worth the extra effort?

Reply:
Maybe not, I was just worried that "materializing" all the events into a `Vec` may be problematic.

Reply:
They are just being moved. The "draining" happens twice over the lifetime of a node, as in only when we change from initializer to joiner, and from joiner to validator. I think it is fairly unproblematic, and we are aiming to replace this in the mid-term.